# Populasi, sampel dan observasi/pengamatan
Observasi: Adalah suatu unit yang diukur dengan data. Beberapa contoh diantaranya adalah:

Siswa
Warga negara
Hewan
Kendaraan
Populasi: Adalah koleksi dari keseluruhan observasi. Beberapa contoh diantaranya adalah:

Semua siswa yang ada di sekolah
Semua warga negara Indonesia,
Semua hewan yang ada di hutan lindung,
Semua kendaraan di Jakarta
Sampel: Adalah sub koleksi dari populasi. Beberapa contoh diantaranya adalah:

5 siswa dari masing-masing kelas di suatu sekolah
100 warga negara Indonesia yang diambil dari beberapa wilayah
Beberapa species hewan tertentu di suatu hutan lindung
3 jenis kendaraan di Jakarta

# Statistik dan Parameter
Parameter adalah penjelasan atas populasi sedangkan statistik hanya menjelaskan sampel dari populasi.

# Data Kualitatif dan Data Kuantitatif
Terdapat dua kategori data yang terdapat pada populasi atau sampel, yaitu data kualitatif dan data kuantitatif.

Data kualitatif adalah data yang diperoleh dari mengkategorikan atau menjelaskan suatu atribut dari populasi atau sampel. Biasanya disebut juga sebagai data kategorik. Beberapa contoh data kualitatif misalnya warna rambut, golongan darah, nama jalan, nama produk yang digunakan dan lain sebagainya.Biasanya data kualitatif selalu disebutkan dalam bentuk kata ataupun simbol.

Data kuantitatif adalah data yang diperoleh dari ukuran atau hitungan di suatu populasi atau sampel. Data kuantitatif selalu berbentuk angka. Data seperti gaji, berat badan, populasi di suatu negara, dan jumlah pelanggan yang dimiliki suatu e-commerce termasuk data kuantitatif. Data kuantitatif sendiri dapat dibagi menjadi dua yaitu data diskrit dan data kontinu.

Data yang diperoleh dari hasil perhitungan adalah data diskrit: jumlah pelanggan, jumlah produk, dan jumlah telefon yang diterima oleh customer service per harinya adalah beberapa contoh data diskrit.

Data yang diperoleh dari hasil perhitungan namun dapat memuat rasio, desimal, atau bilangan irasional adalah data kontinu: berat badan, tinggi badan, waktu, dan gaji adalah beberapa contoh data kontinu.

# Skala Pengukuran - Data Kategorikal
Ada beberapa skala pengukuran sebagaimana ada beberapa cara kita dapat mengelompokkan objek yang ingin kita ukur. Beberapa tingkatan tersebut adalah:
- Skala Nominal
Skala nominal adalah skala yang digunakan untuk mengkategorikan suatu objek pengamatan dengan objek pengamatan lainnya. Sebagai contoh yang termasuk skala nominal seperti gender, kategori barang, ras, status pernikahan dan lain sebagainya

- Skala Ordinal
Skala ordinal adalah skala yang digunakan untuk mengurutkan suatu objek pengamatan dimana suatu titik pengamatan memiliki nilai yang lebih rendah atau tinggi dibanding nilai lainnya. Sebagai contoh yang termasuk skala nominal adalah kelas, jabatan, tingkat pendidikan, dan sebagainya

# Skala Pengukuran - Data Numerikal
- Skala Interval
Skala interval adalah skala yang digunakan untuk tidak hanya untuk mengklasifikasikan maupun memberikan tingkatan pada suatu titik pengamatan, namun kita dapat mengukur seberapa besar nilai antara suatu titik pengamatan dengan titik pengamatan lainnya. Beberapa contoh yang termasuk skala interval diantaranya adalah suhu tubuh dan jarak

- Skala Rasio
Skala rasio memiliki kemiripan dengan skala interval, perbedaannya terletak pada nilai 0 pada skala rasio. Berbeda dengan skala interval yang tidak memiliki nilai 0 yang tidak pasti.

# Package Statistika di Python
Untuk melakukan perhitungan statistika, kita dapat menggunakan beberapa package atau library berikut yang tersedia di Python. Diantaranya adalah:

1. numpy: 
digunakan untuk melakukan analisa data numerik dan perhitungan berbasis vektor atau matriks
2. pandas: 
digunakan untuk melakukan pengolahan data tabular
3. matplotlib: 
digunakan untuk melakukan ploting atau penggambaran grafik, dapat digunakan sebagai alat bantu dalam analisa data
4. statsmodels: 
digunakan untuk melakukan uji hipotesa, eksplorasi data maupun pemodelan statistika
5. scipy: 
digunakan untuk melakukan uji statistika, juga dapat digunakan untuk melakukan pemodelan statistika

Sebenarnya masih banyak lagi library python untuk pengolahan data statistika, namun untuk materi ini cukup 4 atau 5 library di atas saja yang akan kita gunakan.

# Pengenalan Numpy dan Pandas
Kedua library ini, numpy dan pandas, adalah library yang umum digunakan untuk melakukan pengolahan data, mulai dari membaca data dalam format tertentu (csv, xlsx, xls, tsv) atau dari sumber tertentu (RDBMS, No-SQL), melakukan inspeksi (mengecek data yang hilang, inkonsistensi pada data) dan pengolahan data (transformasi nilai, encoding, normalisasi) sampai visualisasi data untuk membuat laporan atau mempersiapkan data untuk membuat model.

Numpy adalah library yang biasanya digunakan untuk manipulasi array atau vektor. Perhitungan yang melibatkan operasi pada objek berbentuk matriks, vektor atau bahkan multidimensi vektor (misalnya data gambar dengan skema warna RGB) dapat kita lakukan dengan menggunakan numpy.

Pandas adalah library yang biasanya digunakan untuk analisa data atau biasa disebut sebagai data wrangling. Biasanya data yang diolah oleh pandas berbentuk tabular atau tabel layaknya spreadsheet di Excel. Pandas menggunakan numpy sebagai back-end sehingga beberapa fungsi atau method dari numpy dapat digunakan di objek pandas.

# Inspeksi Data
Untuk melihat keseluruhan data cukup memanggil nama variabelnya dengan fungsi

In [1]:
import pandas as pd
raw_data = pd.read_csv("https://storage.googleapis.com/dqlab-dataset/dataset_statistic.csv", sep=';')

print (raw_data)
# melihat 10 data pada baris pertama
print (raw_data.head(10))

# melihat 5 data pada baris terakhir
print (raw_data.tail())

    ID Pelanggan     Nama  Jenis Kelamin  Pendapatan Produk   Harga  Jumlah   \
0              1     Arif              1      600000      A  100000        4   
1              2     Dian              2     1200000      D  250000        4   
2              3    Dinda              2      950000      D  250000        3   
3              4    Fajar              1      400000      A  100000        2   
4              5      Ika              2     1200000      D  250000        4   
5              6    Ilham              1      800000      B  150000        4   
6              7    Indra              1      950000      B  150000        5   
7              8  Kartika              2     1100000      E  300000        3   
8              9  Lestari              2      800000      E  300000        2   
9             10      Lia              2     1700000      E  300000        5   
10            11    Maria              2      600000      A  100000        4   
11            12     Maya              2

# Metode Shape
Berapa banyak baris data dan fitur yang dataset punya?

Untuk melihat dimensi dari data kita dapat menggunakan method shape :

In [2]:
import pandas as pd
raw_data = pd.read_csv("https://storage.googleapis.com/dqlab-dataset/dataset_statistic.csv", sep=';')

# melihat dimensi dari raw_data
print (raw_data.shape)

# mengambil jumlah data
print (raw_data.shape[0])

(20, 9)
20


# Melihat Kolom Dalam Dataset
Kolom apa saja yang terdapat dalam dataset?

Untuk melihat kolom apa saja yang terdapat pada dataset cukup menggunakan method columns: 

In [3]:
import pandas as pd
raw_data = pd.read_csv("https://storage.googleapis.com/dqlab-dataset/dataset_statistic.csv", sep=';')

print (raw_data)
print (raw_data.columns)

    ID Pelanggan     Nama  Jenis Kelamin  Pendapatan Produk   Harga  Jumlah   \
0              1     Arif              1      600000      A  100000        4   
1              2     Dian              2     1200000      D  250000        4   
2              3    Dinda              2      950000      D  250000        3   
3              4    Fajar              1      400000      A  100000        2   
4              5      Ika              2     1200000      D  250000        4   
5              6    Ilham              1      800000      B  150000        4   
6              7    Indra              1      950000      B  150000        5   
7              8  Kartika              2     1100000      E  300000        3   
8              9  Lestari              2      800000      E  300000        2   
9             10      Lia              2     1700000      E  300000        5   
10            11    Maria              2      600000      A  100000        4   
11            12     Maya              2

# Metode Isna
Ada berapa banyak data yang hilang dari dataset?

Untuk melihat data dari dataset bisa menggunakan method isna: 

In [4]:
import pandas as pd
raw_data = pd.read_csv("https://storage.googleapis.com/dqlab-dataset/dataset_statistic.csv", sep=';')

print (raw_data.isna())
print (raw_data.isna().sum())

    ID Pelanggan   Nama  Jenis Kelamin  Pendapatan  Produk  Harga  Jumlah   \
0          False  False          False       False   False  False    False   
1          False  False          False       False   False  False    False   
2          False  False          False       False   False  False    False   
3          False  False          False       False   False  False    False   
4          False  False          False       False   False  False    False   
5          False  False          False       False   False  False    False   
6          False  False          False       False   False  False    False   
7          False  False          False       False   False  False    False   
8          False  False          False       False   False  False    False   
9          False  False          False       False   False  False    False   
10         False  False          False       False   False  False    False   
11         False  False          False       False   False  Fals

# Metode Describe
Untuk bisa melihat ringkasan dari data misalnya rerata, jumlah, nilai maksimum-minimum dan ukuran lainnya, kita dapat menggunakan method .describe()

In [5]:
import pandas as pd
raw_data = pd.read_csv("https://storage.googleapis.com/dqlab-dataset/dataset_statistic.csv", sep=';')

print (raw_data.describe())

# Mencari nilai maksimum dari tiap kolom
raw_data.max()
 
# Mencari nilai maksimum dari kolom 'Harga'
raw_data['Harga'].max()
 
# Mencari nilai minimum dari kolom 'Harga'
raw_data['Harga'].min()

       ID Pelanggan  Jenis Kelamin    Pendapatan          Harga    Jumlah   \
count      20.00000      20.000000  2.000000e+01      20.000000  20.000000   
mean       10.50000       1.700000  1.160000e+06  197500.000000   3.650000   
std         5.91608       0.470162  1.282842e+06   69726.910912   1.182103   
min         1.00000       1.000000  4.000000e+05  100000.000000   1.000000   
25%         5.75000       1.000000  6.750000e+05  150000.000000   3.000000   
50%        10.50000       2.000000  8.750000e+05  200000.000000   4.000000   
75%        15.25000       2.000000  1.125000e+06  250000.000000   4.250000   
max        20.00000       2.000000  6.450000e+06  300000.000000   5.000000   

              Total  Tingkat Kepuasan  
count  2.000000e+01         20.000000  
mean   7.100000e+05          2.050000  
std    3.338768e+05          0.887041  
min    2.000000e+05          1.000000  
25%    4.750000e+05          1.000000  
50%    6.750000e+05          2.000000  
75%    9.250000e+

100000

# Metode Sum
Jumlah dari semua nilai pada kolom dengan method .sum()

In [6]:
import pandas as pd
raw_data = pd.read_csv("https://storage.googleapis.com/dqlab-dataset/dataset_statistic.csv", sep=';')

# menghitung jumlah dari semua kolom
print (raw_data.sum())
 
# menghitung jumlah dari semua kolom bertipe data numerik saja
raw_data.sum(numeric_only=True)
 
# menghitung jumlah dari kolom 'Harga' dan 'Pendapatan'
raw_data[['Harga', 'Pendapatan']].sum()

ID Pelanggan                                                      210
Nama                ArifDianDindaFajarIkaIlhamIndraKartikaLestariL...
Jenis Kelamin                                                      34
Pendapatan                                                   23200000
Produk                                           ADDADBBEEEABCDCBCDAC
Harga                                                         3950000
Jumlah                                                             73
Total                                                        14200000
Tingkat Kepuasan                                                   41
dtype: object


Harga          3950000
Pendapatan    23200000
dtype: int64

# Manipulasi Dataframe - Memilih Kolom dan Baris
Terkadang kita hanya ingin melakukan analisa sebagian kecil dari data atau hanya beberapa kolom saja, untuk itu kita dapat melakukan slice-and-dice pada data yang kita punya.

Untuk memilih kolom untuk dianalisa, kita dapat memanggil objek dataframe yang sudah dibuat dan menggunakan tanda ['nama_kolom'] untuk memilih satu kolom atau [['nama_kolom_1', 'nama_kolom_2', ..., 'nama_kolom_n']] untuk memilih lebih dari 1 kolom

In [7]:
import pandas as pd
raw_data = pd.read_csv("https://storage.googleapis.com/dqlab-dataset/dataset_statistic.csv", sep=';')

# Memilih kolom 'Pendapatan' saja
print (raw_data['Pendapatan'])
 
# Memilih kolom 'Jenis Kelamin' dan 'Pendapatan'
print (raw_data[['Jenis Kelamin', 'Pendapatan']])

0      600000
1     1200000
2      950000
3      400000
4     1200000
5      800000
6      950000
7     1100000
8      800000
9     1700000
10     600000
11     950000
12     400000
13    6450000
14    1000000
15     800000
16    1200000
17     700000
18     600000
19     800000
Name: Pendapatan, dtype: int64
    Jenis Kelamin  Pendapatan
0               1      600000
1               2     1200000
2               2      950000
3               1      400000
4               2     1200000
5               1      800000
6               1      950000
7               2     1100000
8               2      800000
9               2     1700000
10              2      600000
11              2      950000
12              2      400000
13              2     6450000
14              2     1000000
15              2      800000
16              1     1200000
17              2      700000
18              2      600000
19              1      800000


# Metode Loc
Untuk memilih baris, kita dapat menggunakan [m:n] untuk mengambil baris data ke-m sampai ke-(n-1) atau bisa juga menggunakan method loc untuk pemilihan baris yang lebih spesifik.

In [8]:
import pandas as pd
raw_data = pd.read_csv("https://storage.googleapis.com/dqlab-dataset/dataset_statistic.csv", sep=';')

# mengambil data dari baris ke-0 sampai baris ke-(10-1) atau baris ke-9
print (raw_data[:10])
 
# mengambil data dari baris ke-3 sampai baris ke-(5-1) atau baris ke-4
print (raw_data[3:5])
 
# mengambil data pada baris ke-1, ke-3 dan ke-10
print (raw_data.loc[[1,3,10]])

# Mengambil kolom 'Jenis Kelamin' dan 'Pendapatan' dan ambil baris ke-1 sampai ke-9
print (raw_data[['Jenis Kelamin', 'Pendapatan']][1:10])
 
# Mengambil kolom 'Harga' dan 'Tingkat Kepuasan' dan ambil baris ke-1, ke-10 dan ke-15
print (raw_data[['Harga', 'Tingkat Kepuasan']].loc[[1,10,15]])

   ID Pelanggan     Nama  Jenis Kelamin  Pendapatan Produk   Harga  Jumlah   \
0             1     Arif              1      600000      A  100000        4   
1             2     Dian              2     1200000      D  250000        4   
2             3    Dinda              2      950000      D  250000        3   
3             4    Fajar              1      400000      A  100000        2   
4             5      Ika              2     1200000      D  250000        4   
5             6    Ilham              1      800000      B  150000        4   
6             7    Indra              1      950000      B  150000        5   
7             8  Kartika              2     1100000      E  300000        3   
8             9  Lestari              2      800000      E  300000        2   
9            10      Lia              2     1700000      E  300000        5   

     Total  Tingkat Kepuasan  
0   400000                 2  
1  1000000                 2  
2   750000                 3  
3   20

# Ukuran Pusat (Measures of Central Tendency)
Ukuran pusat (measures of central tendency) adalah statistika deskriptif yang dapat digunakan untuk membantu kita mengidentifikasi kasus-kasus tipikal di dalam sebuah sampel atau populasi

Terdapat beberapa jenis ukuran pusat yang dapat digunakan untuk menganalisa data, diantaranya:

Mean atau rerata
Median
Modus

# Rata-rata (Mean)
Rata-rata atau mean adalah salah satu ukuran pusat yang nilainya diperoleh dengan cara menjumlahkan semua nilai titik data yang ada lalu dibagi oleh jumlah data.

In [9]:
import numpy as np
import pandas as pd
raw_data = pd.read_csv("https://storage.googleapis.com/dqlab-dataset/dataset_statistic.csv", sep=';')

# mengambil hanya data untuk produk 'A'
produk_A = raw_data[raw_data['Produk']=='A']
 
# menghitung rerata pendapatan menggunakan method .mean pada objek pandas DataFrame
print (produk_A['Pendapatan'].mean())
 
# menghitung rerata pendapatan menggunakan method .mean pada objek pandas DataFrame dengan numpy
print (np.mean(produk_A['Pendapatan']))

550000.0
550000.0


# Median
Median adalah salah satu ukuran pusat yang nilainya terletak di tengah titik data. Sebagai gambaran, jika kita memiliki titik data bernilai 1, 2, 3, 4, 4, 5, 6 maka median dari sekumpulan titik data tersebut adalah 4. 

In [10]:
import numpy as np
import pandas as pd
raw_data = pd.read_csv("https://storage.googleapis.com/dqlab-dataset/dataset_statistic.csv", sep=';')
produk_A = raw_data[raw_data['Produk'] == 'A']

print (raw_data)
# Hitung median dari pendapatan menggunakan pandas
print (produk_A['Pendapatan'].median())

# Hitung median dari pendapatan menggunakan numpy
print (np.median(produk_A['Pendapatan']))

    ID Pelanggan     Nama  Jenis Kelamin  Pendapatan Produk   Harga  Jumlah   \
0              1     Arif              1      600000      A  100000        4   
1              2     Dian              2     1200000      D  250000        4   
2              3    Dinda              2      950000      D  250000        3   
3              4    Fajar              1      400000      A  100000        2   
4              5      Ika              2     1200000      D  250000        4   
5              6    Ilham              1      800000      B  150000        4   
6              7    Indra              1      950000      B  150000        5   
7              8  Kartika              2     1100000      E  300000        3   
8              9  Lestari              2      800000      E  300000        2   
9             10      Lia              2     1700000      E  300000        5   
10            11    Maria              2      600000      A  100000        4   
11            12     Maya              2

# Modus
Modus didefinisikan sebagai data yang memiliki frekuensi kemunculan terbanyak/terbesar. Sebagai contoh, jika terdapat titik data seperti berikut: 1, 1, 1, 1, 2, 3, 3, 4 maka modus dari data tersebut adalah 1 karena 1 muncul sebanyak 4 kali, lebih banyak dibanding titik data lainnya.

Kita dapat menggunakan method .value_counts()

In [11]:
import pandas as pd
raw_data = pd.read_csv("https://storage.googleapis.com/dqlab-dataset/dataset_statistic.csv", sep=';')

# Melihat jumlah dari masing-masing produk
print (raw_data['Produk'].value_counts())


D    5
A    4
B    4
C    4
E    3
Name: Produk, dtype: int64


# Kuantil
Kuantil adalah nilai-nilai data yang membagi data yang telah diurutkan sebelumnya menjadi beberapa bagian yang sama besar ukurannya. Beberapa ukuran fraktil ini diantaranya adalah:

Kuartil: Adalah kuantil yang membagi data menjadi empat bagian sama besar. Kuartil ke-2 dari adalah median dari data tersebut.
Desil: Adalah kuantil yang membagi data menjadi 10 bagian sama besar.
Persentil: Adalah kuantil yang membagi data menjadi 100 bagian sama besar.
Untuk mencari fraktil dari data, kita dapat menggunakan method .quantile

In [12]:
import numpy as np
import pandas as pd
raw_data = pd.read_csv("https://storage.googleapis.com/dqlab-dataset/dataset_statistic.csv", sep=';')

# mencari median atau 50% dari data menggunakan pandas
print (raw_data['Pendapatan'].quantile(q = 0.5))

# mencari median atau 50% dari data menggunakan pandas
print (np.quantile(raw_data['Pendapatan'], q=0.5))

875000.0
875000.0


# Agregasi Data dengan method .agg()
Ada kalanya kita ingin menghitung sekaligus beberapa ukuran, misalnya menghitung nilai mean sekaligus menghitung nilai median. Kita dapat melakukan kedua hal tersebut dengan menggunakan method .agg() pada objek pandas DataFrame

In [13]:
import numpy as np
import pandas as pd
raw_data = pd.read_csv("https://storage.googleapis.com/dqlab-dataset/dataset_statistic.csv", sep=';')

# menghitung rerata dan median usia (age) dan insulin (insu)
print (raw_data[['Pendapatan', 'Harga']].agg([np.mean, np.median]))

# menghitung rerata dan median Pendapatan dan Harga dari tiap produk
print (raw_data[['Pendapatan', 'Harga', 'Produk']].groupby('Produk').agg([np.mean, np.median]))

        Pendapatan     Harga
mean     1160000.0  197500.0
median    875000.0  200000.0
       Pendapatan            Harga        
             mean   median    mean  median
Produk                                    
A          550000   600000  100000  100000
B          875000   875000  150000  150000
C          850000   900000  200000  200000
D         2100000  1200000  250000  250000
E         1200000  1100000  300000  300000


# Ukuran Sebaran (Measures of Dispersion)
Ukuran sebaran (measure of dispersion) adalah statistika deskriptif yang digunakan untuk membantu kita memahami sebaran titik data di dalam sebuah sampel ataupun populasi.

Terdapat beberapa ukuran sebaran yang biasanya digunakan tergantung pada jenis atau tipe datanya, yaitu:

1. Tipe Data Nominal dan Ordinal
- Proporsi Kategori (Categorical Proportion)
- Persen Proporsi (Percent Proportion)
- Rasio Variasi (Variation Ratio)
2. Tipe Data Interval dan Rasio
- Rentang (Range)
- Variansi (Variance)
- Deviasi Baku (Standard Deviation)

# Proporsi Kategori
Proporsi kategori adalah ukuran sebaran yang paling sederhana dari ukuran sebaran pada data nomisal dan ordinal.

In [14]:
import pandas as pd
raw_data = pd.read_csv("https://storage.googleapis.com/dqlab-dataset/dataset_statistic.csv", sep=';')

# cari proporsi tiap Produk
print (raw_data['Produk'].value_counts()/raw_data.shape[0])

D    0.25
A    0.20
B    0.20
C    0.20
E    0.15
Name: Produk, dtype: float64


# Sebaran pada Data Interval dan Rasio
Rentang (range)
Rentang adalah jarak antara nilai maksimum dengan nilai minimum. Semakin besar jarang antara nilai maksimum dan minimum semakin besar pula sebaran datanya. Secara matematis dapat dituliskan sebagai berikut:

Range=max(X)−min(X)

In [15]:

import pandas as pd
raw_data = pd.read_csv("https://storage.googleapis.com/dqlab-dataset/dataset_statistic.csv", sep=';')

# Cari nilai rentang dari kolom 'Pendapatan'
print (raw_data['Pendapatan'].max() - raw_data['Pendapatan'].min())

6050000


# Variansi
Variansi adalah ukuran sebaran pusat yang diperoleh dengan cara menghitung jarak antara tiap titik data pada sampel atau populasi dengan nilai mean. Untuk menghitung variansi kita dapat menggunakan method .var() dari numpy maupun pandas.

pandas menggunakan variansi sampel sedangkan numpy menggunakan variansi populasi. Untuk memperoleh hasil yang sama kita dapat menggunakan parameter ddof=0 pada method .var() untuk memperoleh variansi populasi.

In [16]:

import numpy as np
import pandas as pd
raw_data = pd.read_csv("https://storage.googleapis.com/dqlab-dataset/dataset_statistic.csv", sep=';')

# menghitung variansi umur menggunakan method .var() dari pandas
print (raw_data['Pendapatan'].var())

# menghitung variansi umur menggunakan method .var() dari numpy
print (np.var(raw_data['Pendapatan']))

# mengatur variansi populasi dengan method `.var()` dari pandas
print (raw_data['Pendapatan'].var(ddof = 0))

1645684210526.3157
1563400000000.0
1563400000000.0


# Deviasi Baku (Standard Deviation)
Deviasi baku adalah ukuran sebaran pusat yang diperoleh dengan cara menarik akar kuadrat dari hasil perhitungan variansi. Hal ini dilakukan karena nilai variansi umumnya memiliki nilai yang lebih besar daripada nilai aslinya sebagai efek dari pengkuadratan dan ini menjadikan variansi sulit untuk diinterpretasikan.

menghitung deviasi baku menggunakan method .std dari numpy maupun pandas.

In [17]:
import numpy as np
import pandas as pd
raw_data = pd.read_csv("https://storage.googleapis.com/dqlab-dataset/dataset_statistic.csv", sep=';')

# menghitung deviasi baku sampel pendapatan menggunakan method std() dari pandas
print (raw_data['Pendapatan'].std())

# menghitung deviasi baku pendapatan menggunakan method std() dari numpy
print (np.std(raw_data['Pendapatan'], ddof = 1))

1282842.2391417876
1282842.2391417876


# Korelasi
Korelasi adalah salah satu metode statistika yang dapat digunakan untuk mengukur seberapa besar hubungan antara satu variabel dengan variabel lainnya. Sebagai contoh, misalnya mencari hubungan antara tinggi badan dengan berat badan, mencari hubungan antara gender dengan penghasilan dan masih banyak aplikasi penggunaan korelasi.

Terdapat beberapa metode yang dapat digunakan untuk menghitung korelasi antara sepasang variabel tergantung tipe dari sepasang variabel tersebut. Diantaranya adalah:

1. Korelasi Pearson
2. Korelasi Spearman
3. Korelasi Kendall

# Korelasi Pearson
Korelasi Pearson atau sering juga disebut sebagai Pearson's product moment correlation adalah pengukuran korelasi parametrik yang menghasilkan koefisien korelasi. Koefisien korelasi ini dapat digunakan untuk mengukur kekuatan hubungan atau asosiasi linier antara dua variabel. Artinya jika hubungan kedua variabel tidak linier maka koefisien korelasi Pearson tidak dapat digunakan untuk mengukur kekuatan hubungan antara kedua variabel.

Selain itu nilai dari koefisien pearson dapat digunakan untuk mengukur arah dari hubungan tersebut: positif atau negatif. Hubungan antar variabel dikatakan positif jika nilai salah satu variabel naik maka nilai variabel lainnya juga naik. Sebaliknya, hubungan antar variabel dikatakan negatif jika nilai salah satu variabel naik maka nilai variabelnya turun. Gambar berikut dapat menjelaskan maksud dari kekuatan dan arah dari korelasi antar kedua variabel.

Beberapa asumsi yang harus dipenuhi untuk menggunakan korelasi Pearson diantaranya adalah:

1. Nilai berskala interval/rasio
2. Terdapat hubungan linier antara kedua variabel, yaitu kita dapat menggambarkan hubungan kedua variabel sebagai garis lurus.
3. Kedua variabel berdistribusi normal
4. Homoskedastis, atau data berdistribusi seragam dalam garis regresi

# Korelasi Spearman
Korelasi Spearman atau sering juga disebut sebagai Spearman's rank correlation adalah pengukuran korelasi non-parametrik. Artinya kita mencoba mengukur hubungan antara kedua variabel tanpa menghiraukan asumsi seperti distribusi dari kedua variabel dan asumsi lainnya. Secara kriteria memiliki kemiripan dengan korelasi Pearson walau korelasi Spearman bisa digunakan untuk data bertipe ordinal. Perbedaannya hanya terletak pada pengubahan data dalam bentuk ranking sebelum menghitung nilai korelasinya.

# Korelasi Kendall
Korelasi Kendall atau sering juga disebut juga sebagai Kendall's tank correlation atau korelasi Tau (τ) adalah pengukuran korelasi non-parametrik.

# Menghitung Korelasi
Untuk menghitung korelasi antara kedua variabel, kita dapat menggunakan method .corr() dari pandas

hanya variabel bertipe numerik saja yang dihitung nilai korelasinya. Selain itu secara default method .corr() memiliki parameter method bernilai pearson sehingga nilai korelasi di atas adalah korelasi Pearson. Untuk menggantinya menjadi korelasi Kendall atau korelasi Spearman kita cukup menambahkan parameter method='kendall' atau method='spearman' pada method .corr()

In [18]:

import pandas as pd
raw_data = pd.read_csv("https://storage.googleapis.com/dqlab-dataset/dataset_statistic.csv", sep=';')

# menghitung korelasi dari setiap pasang variabel pada raw_data
print (raw_data.corr())

# mencari korelasi 'kendall' untuk tiap pasang variabel
print (raw_data.corr(method='kendall'))

# mencari korelasi 'spearman' untuk tiap pasang variabel
print (raw_data.corr(method='spearman')) 

                  ID Pelanggan  Jenis Kelamin  Pendapatan     Harga   Jumlah   \
ID Pelanggan          1.000000       0.151375    0.110958 -0.028707  0.011289   
Jenis Kelamin         0.151375       1.000000    0.192849  0.457555 -0.104168   
Pendapatan            0.110958       0.192849    1.000000  0.322443  0.399825   
Harga                -0.028707       0.457555    0.322443  1.000000 -0.138883   
Jumlah                0.011289      -0.104168    0.399825 -0.138883  1.000000   
Total                -0.039968       0.238051    0.592044  0.645455  0.636097   
Tingkat Kepuasan     -0.245717      -0.088339   -0.312663 -0.380798  0.017568   

                     Total  Tingkat Kepuasan  
ID Pelanggan     -0.039968         -0.245717  
Jenis Kelamin     0.238051         -0.088339  
Pendapatan        0.592044         -0.312663  
Harga             0.645455         -0.380798  
Jumlah            0.636097          0.017568  
Total             1.000000         -0.268345  
Tingkat Kepuasan -0.26

# Interpretasi Nilai Korelasi dan Kaitannya dengan Kausalitas
beberapa pasang variabel memiliki nilai korelasi yang positif maupun negatif. Selain itu terdapat nilai yang sangat kecil hingga mendekati nol, namun ada juga yang cukup besar diatas 0.5, misalnya korelasi antara 'Jumlah' dan 'Pendapatan'. Besar kecilnya suatu nilai korelasi dari sepasang variabel menandakan seberapa kuat hubungan linier antara kedua variabel tersebut.

hubungan antara variabel 'Total' dan 'Jumlah' hubungannya sedang dan positif (r = 0.636097) sedangkan hubungan antara 'Pendapatan' dan 'Tingkat Kepuasan' hubungannya lemah dan negatif (r = -0.088339).

Namun, perlu diperhatikan bahwa walaupun kita dapat menilai kuat hubungan antara kedua variabel, namun kita tidak bisa menentukan arah dari hubugan tersebut. Sebagai contoh, walaupun antara variabel 'Total' dan 'Jumlah' memiliki hubungan yang sedang, tidak berarti kita bisa mengetahui apakah kenaikan nilai 'Total' berefek positif terhadap 'Jumlah' atau sebaliknya.